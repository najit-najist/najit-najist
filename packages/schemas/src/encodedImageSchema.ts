import { z } from 'zod';

import { isFileBase64 } from './utils/isFileBase64';

export const IMAGE_FILE_REGEX = /image\/(jp(e)?g|webp|gif|png|svg\+xml)/;

export const encodedImageSchema = z.string().refine((input) => {
  if (input.startsWith('data:')) {
    return isFileBase64(input, IMAGE_FILE_REGEX);
  }

  return true;
});
