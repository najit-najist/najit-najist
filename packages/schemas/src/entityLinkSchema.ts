import { z } from 'zod';

export const entityLinkSchema = z.object(
  {
    id: z.number({ required_error: 'Požadováno' }).min(1, 'Požadováno'),
  },
  { required_error: 'Požadováno' },
);

export type EntityLink = z.infer<typeof entityLinkSchema>;
