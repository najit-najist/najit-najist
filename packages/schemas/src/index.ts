export * from './models';
export * from './utils/getMinimumPickupTimeForDate';
export * from './utils/isFileBase64';
export * from './utils/splitBase64Url';

export * from './encodedImageSchema';
export * from './entityLinkSchema';
export * from './nonEmptyStringSchema';
export * from './packetaMetadataSchema';
export * from './pickupTimeSchema';
export * from './slugSchema';
export * from './streetNameSchema';
export * from './stringOrDateToDateSchema';
export * from './subscribeToNewsletterInputSchema';
export * from './telephoneNumberInputSchema';
export * from './userCartCheckoutInputSchema';
