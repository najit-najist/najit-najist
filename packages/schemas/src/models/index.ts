export * from './municipalitySchema';
export * from './telephoneNumberSchema';
export * from './userAddressSchema';
export * from './userSchema';
