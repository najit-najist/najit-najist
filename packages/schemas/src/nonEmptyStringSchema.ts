import { z } from 'zod';

export const nonEmptyStringSchema = z
  .string({ required_error: 'Požadovaná hodnota' })
  .trim()
  .min(1, 'Požadovaná hodnota');
