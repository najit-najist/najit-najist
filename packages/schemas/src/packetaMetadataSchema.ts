import { PacketaPickupPointType } from '@najit-najist/packeta';
import { PacketaPointError } from '@najit-najist/packeta';
import { z } from 'zod';

export const packetaMetadataSchema = z.object(
  {
    id: z.string(),
    pickupPointType: z.literal(PacketaPickupPointType.INTERNAL, {
      invalid_type_error: 'Tento typ odběrného místa nepodporujeme',
    }),
    error: z
      .nativeEnum(PacketaPointError)
      .nullish()
      .refine(
        (error) => !error,
        'Vybrané místo není dostupné. Vyberete prosím jiné',
      ),
    // carrierId: z.string().nullish(),
    // carrierPickupPointId: z.string().nullish(),
  },
  {
    required_error: 'Dokončete výber odběrného místa',
    invalid_type_error: 'Dokončete výber odběrného místa',
  },
);
