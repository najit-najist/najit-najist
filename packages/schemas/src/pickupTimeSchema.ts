import { z } from 'zod';

import { DEFAULT_TIMEZONE, dayjs } from './internals/dates';

export const pickupTimeSchema = z
  .string({ required_error: 'Vyplňte čas vyzvednutí' })
  .min(1, 'Vyplňte čas vyzvednutí')
  .superRefine((valueAsString, ctx) => {
    const value = dayjs(valueAsString).tz(DEFAULT_TIMEZONE, true);
    if (!value.isValid()) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Vybraný datum a čas musí být ve správném formátu',
        fatal: true,
      });

      return;
    }

    const now = dayjs().tz(DEFAULT_TIMEZONE);
    const diffInMinutes = value.diff(now, 'minutes');

    if (diffInMinutes < 2 * 60) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Datum a čas musí být alespoň dvě hodiny od aktuálního času',
        fatal: true,
      });
    }
  });
