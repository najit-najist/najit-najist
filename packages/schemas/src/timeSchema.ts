import { z } from 'zod';

const timeRegex = /^[0-2]?[0-9]:[0-5]?[0-9]$/;

export type Time = `${number}:${number}`;
export type ZodTimeOptions = {
  max?: Time;
  min?: Time;
};

const MAX_HOUR = 23;
const MAX_MINUTE = 59;

const isValidTimeString = (value: any): value is `${number}:${number}` => {
  if (typeof value !== 'string' || !timeRegex.test(value)) {
    return false;
  }

  const [hours, minutes] = value.split(':');

  return !Number.isNaN(+hours) && !Number.isNaN(+minutes) && +hours <= 23;
};

const timeToNumber = (value: Time): number => Number(value.replace(':', ''));

export const zodTime = (options?: ZodTimeOptions) => {
  let { max = `${MAX_HOUR}:${MAX_MINUTE}`, min = `00:00` } = {
    ...(options ?? {}),
  };

  const minAsNumber = timeToNumber(min);
  const maxAsNumber = timeToNumber(max);

  return z
    .custom<Time>(
      (value) => {
        if (!isValidTimeString(value)) {
          return false;
        }

        const valueAsNumber = timeToNumber(value);

        return minAsNumber <= valueAsNumber && maxAsNumber >= valueAsNumber;
      },
      (value) => {
        let message = 'Čas musí být vyplněn a ve formátu HH:MM';
        if (isValidTimeString(value)) {
          const valueAsNumber = timeToNumber(value);
          if (minAsNumber > valueAsNumber) {
            message = `Čas musí být alespoň od ${min}`;
          } else if (maxAsNumber < valueAsNumber) {
            message = `Čas musí být maximálně do ${min}`;
          }
        }

        return {
          fatal: true,
          message,
        };
      }
    )
    .transform((value) => {
      const [hours, minutes] = value.split(':');

      return `${hours.padStart(2, '0')}:${minutes.padStart(2, '0')}`;
    });
};

export const timeSchema = zodTime();
