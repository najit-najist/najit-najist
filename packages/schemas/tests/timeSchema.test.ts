import { describe, expect, it } from 'vitest';

import { Time, ZodTimeOptions, zodTime } from '../src/timeSchema';

describe('timeSchema', () => {
  it.each(['hellothere', 'bad:00', '00:bad', '..', ':', '--:--'])(
    'should fail on incorrect input %c',
    (variant) => {
      const parsedString = zodTime().safeParse(variant);

      expect(parsedString.success).toBeFalsy();
      if (!parsedString.success) {
        expect(parsedString.error).toMatchSnapshot();
      }
    }
  );

  it.each([
    ['0:1', '00:01'],
    ['00:59', '00:59'],
    ['23:01', '23:01'],
    ['10:59', '10:59'],
  ])('should correctly parse correct time %s', (input, output) => {
    const parsedString = zodTime().safeParse(input);

    if (!parsedString.success) {
      throw new Error('should result in truthy');
    }

    expect(parsedString.data).toBe(output);
  });

  it.each([
    [{ max: '10:30' }, `10:31`],
    [{ min: '10:30' }, `9:00`],
  ] as Array<[ZodTimeOptions, Time]>)(
    'it correctly throws on incorrect time',
    (options, input) => {
      const parsedString = zodTime(options).safeParse(input);

      expect(parsedString.success).toBeFalsy();
      expect((parsedString as any).error).toMatchSnapshot();
    }
  );
});
