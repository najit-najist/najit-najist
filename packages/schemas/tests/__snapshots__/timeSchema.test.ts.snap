// Vitest Snapshot v1

exports[`timeSchema > it correctly throws on incorrect time 1`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být maximálně do 00:00",
    "path": []
  }
]]
`;

exports[`timeSchema > it correctly throws on incorrect time 2`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být alespoň od 10:30",
    "path": []
  }
]]
`;

exports[`timeSchema > should fail on incorrect input  1`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být vyplněn a ve formátu HH:MM",
    "path": []
  }
]]
`;

exports[`timeSchema > should fail on incorrect input  2`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být vyplněn a ve formátu HH:MM",
    "path": []
  }
]]
`;

exports[`timeSchema > should fail on incorrect input  3`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být vyplněn a ve formátu HH:MM",
    "path": []
  }
]]
`;

exports[`timeSchema > should fail on incorrect input  4`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být vyplněn a ve formátu HH:MM",
    "path": []
  }
]]
`;

exports[`timeSchema > should fail on incorrect input  5`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být vyplněn a ve formátu HH:MM",
    "path": []
  }
]]
`;

exports[`timeSchema > should fail on incorrect input  6`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "message": "Čas musí být vyplněn a ve formátu HH:MM",
    "path": []
  }
]]
`;
